ksp_plugin/plugin.cpp

```diff
@@ -149,127 +124,120 @@ Plugin::~Plugin() {
-                                                    parent_index).second;
-  inserted &=
-      hierarchical_initialization_->
-          indices_to_bodies.emplace(celestial_index, unowned_body).second;
-  CHECK(inserted);
-
-  // Record the fingerprints of the parameters to detect if we are in KSP stock.
-  CHECK(celestial_jacobi_keplerian_fingerprints_.insert(
-            FingerprintCelestialJacobiKeplerian(celestial_index,
-                                                parent_index,
-                                                keplerian_elements,
-                                                *unowned_body)).second);
 }
 
 void Plugin::EndInitialization() {
   CHECK(initializing_);
-  if (hierarchical_initialization_) {
-    std::uint64_t system_fingerprint = 0;
-    for (std::uint64_t fingerprint : celestial_jacobi_keplerian_fingerprints_) {
-      system_fingerprint = FingerprintCat2011(system_fingerprint, fingerprint);
-    }
-    LOG(INFO) << "System fingerprint is " << std::hex << system_fingerprint;
+  SolarSystem<Barycentric> solar_system(gravity_model_, initial_state_);
+
+  // If the system was constructed using keplerian elements, it may be the
+  // stock KSP system in which case it needs to be stabilized.
+  if (initial_state_.has_keplerian()) {
+    auto const hierarchical_system = solar_system.MakeHierarchicalSystem();
+    serialization::HierarchicalSystem message;
+    hierarchical_system->WriteToMessage(&message);
+    std::string const serialized_message = message.SerializeAsString();
+    uint64_t const system_fingerprint =
+        Fingerprint2011(serialized_message.c_str(), serialized_message.size());
+    LOG(INFO) << "System fingerprint is " << std::hex << std::uppercase
+              << system_fingerprint;
+
     if (system_fingerprint == ksp_stock_system_fingerprint) {
-      is_ksp_stock_system_ = true;
       LOG(WARNING) << "This appears to be the dreaded KSP stock system!";
-    } else if (system_fingerprint == ksp_fixed_system_fingerprint) {
-      LOG(INFO) << "This is the fixed KSP system, all hail retrobop!";
+      StabilizeKSP(solar_system);
+      auto const hierarchical_system = solar_system.MakeHierarchicalSystem();
+      serialization::HierarchicalSystem message;
+      hierarchical_system->WriteToMessage(&message);
+      std::string const serialized_message = message.SerializeAsString();
+      uint64_t const system_fingerprint = Fingerprint2011(
+          serialized_message.c_str(), serialized_message.size());
+      LOG(INFO) << "System fingerprint after stabilization is " << std::hex
+                << std::uppercase << system_fingerprint;
+      CHECK_EQ(ksp_stabilized_system_fingerprint, system_fingerprint)
+          << "Attempt at stabilizing the KSP system failed!\n"
+          << gravity_model_.DebugString() << "\n"
+          << initial_state_.DebugString();
+      LOG(INFO) << "This is the stabilized KSP system, all hail retrobop!";
+    } else {
```

Comment:
Add an `else if` clause that checks if this is already the retrobop system; this may happen when using a Kopernicus config that's equivalent to `StabilizeKSP` (in fact we should probably check that it is possible to write such a Kopernicus config, otherwise we're doing inconsistent things in `StabilizeKSP`).